README.md
# Laravel-Elasticsearch

An easy way to use the [official Elastic Search client](https://github.com/elastic/elasticsearch-php) in your Laravel or Lumen applications.

> ## ⚠️ This Package Will Be Abandoned
> Due to several factors, including the fact that I no longer use ES,
> I will be ceasing development on this package.  If you are interested in
> taking over this project, please reach out to me here on the Issues board.

[![Build Status](https://github.com/cviebrock/laravel-elasticsearch/workflows/tests/badge.svg?branch=master)](https://github.com/cviebrock/laravel-elasticsearch/actions)
[![Total Downloads](https://poser.pugx.org/cviebrock/laravel-elasticsearch/downloads.png)](https://packagist.org/packages/cviebrock/laravel-elasticsearch)
[![Latest Stable Version](https://poser.pugx.org/cviebrock/laravel-elasticsearch/v/stable.png)](https://packagist.org/packages/cviebrock/laravel-elasticsearch)
[![Latest Stable Version](https://poser.pugx.org/cviebrock/laravel-elasticsearch/v/unstable.png)](https://packagist.org/packages/cviebrock/laravel-elasticsearch)
[![License](https://img.shields.io/packagist/l/cviebrock/laravel-elasticsearch)](LICENSE.md)


- [Installation and Configuration](#installation-and-configuration)
  - [Laravel](#laravel)
    - [Alternative configuration method via .env file](#alternative-configuration-method-via-env-file)
    - [Connecting to AWS Elasticsearch Service](#connecting-to-aws-elasticsearch-service)
  - [Lumen](#lumen)
- [Usage](#usage)
- [Advanced Usage](#advanced-usage)
- [Console Commands](#console-commands)
- [Bugs, Suggestions, Contributions and Support](#bugs-suggestions-contributions-and-support)
- [Copyright and License](#copyright-and-license)



## Installation and Configuration

Install the current version of the `cviebrock/laravel-elasticsearch` package via composer:

```sh
composer require cviebrock/laravel-elasticsearch
```

If you are using ElasticSearch version 5, then install version 2 of this package:

```sh
composer require cviebrock/laravel-elasticsearch:^2
```

### Laravel

The package's service provider will automatically register its service provider.

Publish the configuration file:

```sh
php artisan vendor:publish --provider="Cviebrock\LaravelElasticsearch\ServiceProvider"
```

##### Alternative configuration method via .env file

After you publish the configuration file as suggested above, you may configure ElasticSearch
by adding the following to your application's `.env` file (with appropriate values):
  
```ini
ELASTICSEARCH_HOST=localhost
ELASTICSEARCH_PORT=9200
ELASTICSEARCH_SCHEME=http
ELASTICSEARCH_USER=
ELASTICSEARCH_PASS=
```

If you are logging in via API keys, you will need to fill in these values:

```ini
ELASTICSEARCH_API_ID=
ELASTICSEARCH_API_KEY=
```


##### Connecting to AWS Elasticsearch Service

If you are connecting to ElasticSearch instances on Amazon AWS, then you'll also
need to `composer require aws/aws-sdk-php:^3.80` and add the following to your 
`.env` file:

```ini
AWS_ELASTICSEARCH_ENABLED=true
AWS_REGION=...
AWS_ACCESS_KEY_ID=...
AWS_SECRET_ACCESS_KEY=...
```

If you have to use another authentication method having custom credentials (i.e. `instanceProfile()`), 
you have to publish the configuration file and use the **aws_credentials**:

```php
<?php
// config/elasticsearch.php

$provider = \Aws\Credentials\CredentialProvider::instanceProfile();
$memoizedProvider = \Aws\Credentials\CredentialProvider::memoize($provider);
$credentials = $memoizedProvider()->wait();

...

'hosts' => [
    [
        'host'            => env('ELASTICSEARCH_HOST', 'localhost'),
        // For local development, the default Elasticsearch port is 9200.
        // If you are connecting to an Elasticsearch instance on AWS, you probably want to set this to null
        'port'            => env('ELASTICSEARCH_PORT', 9200),
        'scheme'          => env('ELASTICSEARCH_SCHEME', null),
        'user'            => env('ELASTICSEARCH_USER', null),
        'pass'            => env('ELASTICSEARCH_PASS', null),

        // If you are connecting to an Elasticsearch instance on AWS, you will need these values as well
        'aws'             => env('AWS_ELASTICSEARCH_ENABLED', false),
        'aws_region'      => env('AWS_REGION', ''),
        'aws_key'         => env('AWS_ACCESS_KEY_ID', ''),
        'aws_secret'      => env('AWS_SECRET_ACCESS_KEY', '')
        'aws_credentials' => $credentials
    ],
],
```

If you have a job that runs in supervisor, you have to use the Closure.
This way the credentials will be renewed at runtime.

```php
<?php
// config/elasticsearch.php

$provider = \Aws\Credentials\CredentialProvider::instanceProfile();
$memoizedProvider = \Aws\Credentials\CredentialProvider::memoize($provider);

...

'hosts' => [
    [
        ...
        'aws_credentials' => $memoizedProvider
    ],
],
```

If you are using `php artisan config:cache`, you cannot have the Closure in your config file, call it like this:

```php
<?php
// config/elasticsearch.php

...

'hosts' => [
    [
        ...
        'aws_credentials' => [\Aws\Credentials\CredentialProvider::class, 'defaultProvider'],
    ],
],
```

### Lumen

If you work with Lumen, please register the service provider and configuration in `bootstrap/app.php`:

```php
$app->register(Cviebrock\LaravelElasticsearch\ServiceProvider::class);
$app->configure('elasticsearch');
```

Manually copy the configuration file to your application.



## Usage

The `Elasticsearch` facade is just an entry point into the [ES client](https://github.com/elastic/elasticsearch-php),
so previously you might have used:

```php
use Elasticsearch\ClientBuilder;

$data = [
    'body' => [
        'testField' => 'abc'
    ],
    'index' => 'my_index',
    'type' => 'my_type',
    'id' => 'my_id',
];

$client = ClientBuilder::create()->build();
$return = $client->index($data);
```

You can now replace those last two lines with simply:

```php
use Elasticsearch;

$return = Elasticsearch::index($data);
```

That will run the command on the default connection.  You can run a command on
any connection (see the `defaultConnection` setting and `connections` array in
the configuration file).

```php
$return = Elasticsearch::connection('connectionName')->index($data);
```

Lumen users who wish to use Facades can do so by editing the 
`bootstrap/app.php` file to include the following:

```php
$app->withFacades(true, [
    ...
    Cviebrock\LaravelElasticsearch\Facade::class => 'Elasticsearch',
    ...
]);
```

Lumen users who aren't using facades will need to use dependency injection 
or the application container in order to get the ES service object:

```php
// using injection:
public function handle(\Cviebrock\LaravelElasticsearch\Manager $elasticsearch)
{
    $elasticsearch->ping();
}

// using application container:
$elasticSearch = $this->app('elasticsearch');
```

Of course, dependency injection and the application container work 
for Laravel applications as well.



## Advanced Usage

Because the package is a wrapper around the official Elastic client, you can 
do pretty much anything with this package.  Not only can you perform standard 
CRUD operations, but you can monitor the health of your Elastic cluster programmatically, 
back it up, or make changes to it.  Some of these operations are done through 
"namespaced" commands, which this package happily supports.

To grab statistics for an index:

```php
$stats = Elasticsearch::indices()->stats(['index' => 'my_index']);
$stats = Elasticsearch::nodes()->stats();
$stats = Elasticsearch::cluster()->stats();
```

To create and restore snapshots (read the Elastic docs about creating repository paths and plugins first):

```php
$response = Elasticsearch::snapshots()->create($params);
$response = Elasticsearch::snapshots()->restore($params);
```

To delete whole indices (be careful!):

```php
$response = Elasticsearch::indices()->delete(['index' => 'my_index']);
```

Please remember that this package is a thin wrapper around a large number of very 
sophisticated and well-documented Elastic features.  Information about those features 
and the methods and parameters used to call them can be found in the 
[Elastic documentation](https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/index.html).
Help with using them is available via the [Elastic forums](https://discuss.elastic.co/) 
and on sites like [Stack Overflow](https://stackoverflow.com/questions/tagged/elasticsearch).



## Console commands

This package also provides some useful console commands.

Check if an index exists:

```sh
php artisan laravel-elasticsearch:utils:index-exists <your_elasticsearch_index_name>
```

Create an index:

```sh
php artisan laravel-elasticsearch:utils:index-create <your_elasticsearch_index_name>
```

Delete an index:

```sh
php artisan laravel-elasticsearch:utils:index-delete <your_elasticsearch_index_name>
```

Create or update index mapping:  
_Note: The index mapping file must contain a valid JSON mapping definition as Elasticsearch expects, for example:_

```json
{
    "body": {
        "_source": {
            "enabled": true
        },
        "properties": {
            "id": {
                "type": "keyword"
            },
            "property_1": {
                "type": "text"
            },
            "property_2": {
                "type": "text"
            }
        }
    }
}
```

```sh
php artisan laravel-elasticsearch:utils:index-create-or-update-mapping <your_elasticsearch_index_name> <json_mapping_absolute_file_path>
```

Creates an alias:

```sh
php artisan laravel-elasticsearch:utils:alias-create <your_elasticsearch_index_name> <your_elasticsearch_alias_name>
```

Remove index from an alias:

```sh
php artisan laravel-elasticsearch:utils:alias-remove-index <your_elasticsearch_index_name> <your_elasticsearch_alias_name>
```

Switch index on alias (useful for zero-downtime release of the new index):

```sh
php artisan laravel-elasticsearch:utils:alias-switch-index <your_NEW_elasticsearch_index_name> <your_OLD_elasticsearch_index_name> <your_elasticsearch_alias_name>
```  



## Bugs, Suggestions, Contributions and Support

Thanks to [everyone](https://github.com/cviebrock/laravel-elasticsearch/graphs/contributors)
who has contributed to this project!

Please use [Github](https://github.com/cviebrock/laravel-elasticsearch) for reporting bugs, 
and making comments or suggestions.
 
See [CONTRIBUTING.md](CONTRIBUTING.md) for how to contribute changes.



## Copyright and License

[laravel-elasticsearch](https://github.com/cviebrock/laravel-elasticsearch)
was written by [Colin Viebrock](http://viebrock.ca) and is released under the 
[MIT License](LICENSE.md).

Copyright (c) 2015 Colin Viebrock


CHANGELOG.md
# Changelog

## 9.0.3 - 07-Nov-2022

- Add ability to log in via API keys (#141, thanks @mhdEurFlash123)


## 9.0.2 - 18-Jul-2022

- "Update" `guzzlehttp/psr7` package to version `^1.7|^2.0` (thanks @brunogaspar)


## 9.0.1 - 16-Jul-2022

- Update `guzzlehttp/psr7` package to version `^2.0` (currently 2.4.0) (#138)


## 9.0.0 - 23-Feb-2022

- Support for Laravel 7, 8 and 9, and PHP 7.3–8.1 (as appropriate)


## 8.0.7 - 05-Jan-2022

- Fix `composer.json` for PHP 8 compatability (#133, thanks @adriacanal)


## 8.0.6 - 25-Oct-2021

- Add comment about `hosts.port` needing to be `null` for typical AWS connections (#126, thanks @sp-clariondoor)


## 8.0.5 - 22-Sep-2021

- Fix host assignment issue (#124, thanks @sete391)


## 8.0.4 - 28-Feb-2020

- Support for PHP 8.0 (thanks @okdewit)
  - bumps ES client version to `^7.11`


## 8.0.3 - 28-Dec-2020

- Support using a closure for AWS credentials, allowing the config to be cached (#111, thanks @stasadev)


## 8.0.2 - 02-Dec-2020

- Move automated testing from travis-ci to Github actions
- Clean up some third-party tools and badges
- Ability to create a new index with settings/mappings on `IndexCreateOrUpdateMappingCommand` (#106, thanks zeidanbm)
- Fix for loading JSON mapping files in `IndexCreateOrUpdateMappingCommand` (#104, thanks @danieljaniga)
- Set port for the request (#105, thanks @joshuaeon)
- Better error handling for request statistics (#105, thanks @joshuaeon)


## 8.0.1 - 29-Sep-2020

- Add utility console commands (huge thanks to @nsaliu)


## 8.0.0 - 10-Sep-2020

- Support for Laravel 8.0 (and drop support for earlier versions due to PHP and package conflicts) 
- Added AWS Session Token support (#87, thanks @nsaliu)


## 4.2.1 - 04-Mar-2020

- Support for custom AWS credentials (#80, thanks @giorgiomen)


## 4.2.0 - 04-Mar-2020

- Support for Laravel 7.0 (and drop support for < 6.0)


## 4.1.3 - 09-Feb-2020

- Support for AWS credentials (#76, thanks @giorgiomen)
- Speed up Travis
 

## 4.1.2 - 19-Jan-2020

- Minor fixes (add PHP 7.4 testing)


## 4.1.1 - 13-Sep-2019

- Fix for semantic versioning


## 4.1.0 - 03-Sep-2019

- Support for Laravel 6.0 (and drop support for < 5.8)


## 4.0.0 - 31-Aug-2019

- Bump ES client version to ^7.0 (#68, thanks @Lednerb)


## 3.6.0 - 26-Jun-2019

- Add support for tracers (#65, thanks @luoxiaojun1992)
- Fix typo in README.md (#63, thanks @Harrisonbro)


## 3.5.1 - 29-Apr-2019

- Fix for ElasticSearch 6.5 (#59, thanks @SirNarsh)


## 3.5.0 - 04-Mar-2019

- Support Laravel/Lumen 5.8


## 3.4.0 - 17-Dec-2018

- Re-add Lumen support (thanks @ristedavcevski)
- Fix to AWS host logic (@thanks @tufanbarisyildirim)


## 3.3.0 - 12-Dec-2018

- Support AWS hosts (thanks @Matrix86)
- Add ability to register additional namespaces (thanks @ShaneXie)


## 3.2.0 - 04-Sep-2018

- Support Laravel/Lumen 5.7


## 3.1.1 - 28-Feb-2018

- Support Laravel/Lumen 5.6 (thanks @akira28)


## 3.1.0 - 14-Dec-2017 

- Real fix for Lumen users (thanks @petercoles and @aripekkako)


## 3.0.0 - 29-Nov-2017

- Bump ES client version to ^6.0


## 2.1.0 - 31-Aug-2017

- Laravel 5.5 support, including auto-registration
- Bump ES client version to ^5.3


## 2.0.1 - 14-Jul-2017

* Add alias for Lumen (thanks @matejvelikonja)


## 2.0.0 - 04-May-2017

* Bump ES client version to 5.x
* Allow configuration via `.env` file (thanks @expertcoder)


## 1.3.0 - 17-Feb-2017

* Further fixes for Lumen (thanks @pafelin).


## 1.2.0 - 20-Jun-2016

* Fixed support for Lumen (thanks @JamesB797).
* Fix for omitting connection configuration (thanks @xian13).
* PSR-2 code style changes.
* Cleaned up Lumen service provider, and refactored configuration setting in Factory.


## 1.1.0 - 11-Apr-2016

* Added support for Lumen (thanks @khoatran).


## 1.0.0 - 06-Jan-2016

* First "stable" release.
* Replace `bindShared` with `singleton` in the service provider to make it Laravel 5.2 compatible (#5 thanks @goldlife)


## 0.9.1 - 04-Jun-2015

* Remove index mangling from this package (that functionality belongs elsewhere).


## 0.9.0 - 27-May-2015

* Initial tagged release.


config/elasticsearch.php
<?php

return [

    /**
     * You can specify one of several different connections when building an
     * Elasticsearch client.
     *
     * Here you may specify which of the connections below you wish to use
     * as your default connection when building an client. Of course you may
     * use create several clients at once, each with different configurations.
     */

    'defaultConnection' => 'default',

    /**
     * These are the connection parameters used when building a client.
     */

    'connections' => [

        'default' => [

            /**
             * Hosts
             *
             * This is an array of hosts that the client will connect to. It can be a
             * single host, or an array if you are running a cluster of Elasticsearch
             * instances.
             *
             * This is the only configuration value that is mandatory.
             *
             * Presently using "extended" host configuration method
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_extended_host_configuration
             *
             * There is also the shorter "inline" configuration method available
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_inline_host_configuration
             */

            'hosts' => [
                        env('ELASTICSEARCH_SCHEME', 'http').'://'.env('ELASTICSEARCH_HOST', 'localhost').':'.env('ELASTICSEARCH_PORT', 9200),
                    // For local development, the default Elasticsearch port is 9200.
            ],
            'basicAuthentication' => [
                'user' => env('ELASTICSEARCH_USER', null),
                'pass' => env('ELASTICSEARCH_PASS', null),
            ],
            'cloud_api' => [
                // Alternatively, you can log in via API keys
                'api_id'            => env('ELASTICSEARCH_API_ID', null),
                'api_key'           => env('ELASTICSEARCH_API_KEY', null),
            ],

            /**
             * SSL
             *
             * If your Elasticsearch instance uses an out-dated or self-signed SSL
             * certificate, you will need to pass in the certificate bundle.  This can
             * either be the path to the certificate file (for self-signed certs), or a
             * package like https://github.com/Kdyby/CurlCaBundle.  See the documentation
             * below for all the details.
             *
             * If you are using SSL instances, and the certificates are up-to-date and
             * signed by a public certificate authority, then you can leave this null and
             * just use "https" in the host path(s) above and you should be fine.
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_security.html#_ssl_encryption_2
             */

            'sslVerification' => null,

            /**
             * Logging
             *
             * Logging is handled by passing in an instance of Monolog\Logger (which
             * coincidentally is what Laravel's default logger is).
             *
             * If logging is enabled, you either need to set the path and log level
             * (some defaults are given for you below), or you can use a custom logger by
             * setting 'logObject' to an instance of Psr\Log\LoggerInterface.  In fact,
             * if you just want to use the default Laravel logger, then set 'logObject'
             * to \Log::getMonolog().
             *
             * Note: 'logObject' takes precedent over 'logPath'/'logLevel', so set
             * 'logObject' null if you just want file-based logging to a custom path.
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#enabling_logger
             */

            'logging' => false,

            // If you have an existing instance of Monolog you can use it here.
            // 'logObject' => \Log::getMonolog(),

            'logPath' => storage_path('logs/elasticsearch.log'),

            'logLevel' => Monolog\Logger::INFO,

            /**
             * Retries
             *
             * By default, the client will retry n times, where n = number of nodes in
             * your cluster. If you would like to disable retries, or change the number,
             * you can do so here.
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_set_retries
             */

            'retries' => null,

            /**
             * The remainder of the configuration options can almost always be left
             * as-is unless you have specific reasons to change them.  Refer to the
             * appropriate sections in the Elasticsearch documentation for what each option
             * does and what values it expects.
             */

            /**
             * Sniff On Start
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html
             */

            'sniffOnStart' => false,

            /**
             * HTTP Handler
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_configure_the_http_handler
             * @see http://ringphp.readthedocs.org/en/latest/client_handlers.html
             */

            'httpHandler' => null,

            /**
             * Connection Pool
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_setting_the_connection_pool
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_connection_pool.html
             */

            'connectionPool' => null,

            /**
             * Connection Selector
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_setting_the_connection_selector
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_selectors.html
             */

            'connectionSelector' => null,

            /**
             * Serializer
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_setting_the_serializer
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_serializers.html
             */

            'serializer' => null,

            /**
             * Connection Factory
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_setting_a_custom_connectionfactory
             */

            'connectionFactory' => null,

            /**
             * Endpoint
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/6.0/_configuration.html#_set_the_endpoint_closure
             */

            'endpoint' => null,


            /**
             * Register additional namespaces
             *
             * An array of additional namespaces to register.
             *
             * @example 'namespaces' => [XPack::Security(), XPack::Watcher()]
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/current/ElasticsearchPHP_Endpoints.html#Elasticsearch_ClientBuilderregisterNamespace_registerNamespace
             */
            'namespaces' => [],

            /**
             * Tracer
             *
             * Tracer is handled by passing in a name of the class implements Psr\Log\LoggerInterface.
             *
             * @see https://www.elastic.co/guide/en/elasticsearch/client/php-api/2.0/_configuration.html#_setting_a_custom_connectionfactory
             */
            'tracer' => null,

        ],

    ],

];
